executable/vba.ksy

```diff
@@ -0,0 +1,1209 @@
+              - CodeStart
+              - StartOfCode
+            type: resolved_ptr
+          - id: code_end_ptr
+            -orig-id:
+              - CodeEnd
+              - EndOfCode
+            type: resolved_ptr
+          - id: data_size
+            -orig-id:
+              - dwDataSize
+              - lDataBufferSize
+            type: u4
+          - id: thread_space_ptr
+            -orig-id: ThreadSpace
+            type: resolved_ptr
+          - id: vba_seh_ptr
+            -orig-id:
+              - VbaSeh
+              - VBAExceptionhandler
+            type: resolved_ptr
+          - id: native_code_ptr
+            -orig-id: NativeCode
+            type: resolved_ptr
+          - id: primitive_path
+            -orig-id: wsPrimitivePath
+            type: strz
+            #encoding: utf-16le
+            size: 6
+            if: _root.header.runtime_build < 9782
+          - id: path_information
+            -orig-id: szPathInformation
+            type: str # fucking incorrectly processes strz
+            #pad-right: 0x00
+            encoding: utf-16le
+            size: "0x210 - (_root.header.runtime_build < 9782 ? 6 : 0)"
+          - id: external_table
+            -orig-id: lpExternalTable
+            type: u4
+          - id: external_count
+            -orig-id: dwExternalCount
+            type: u4
+
+        instances:
+          object_table:
+            io: _root._io
+            pos: object_table_ptr.ofs
+            type: object_table
+
+      project_data_2:
+        -orig-id: ProjectData2
+        seq:
+          - id: heap_link_ptr
+            -orig-id: lpHeapLink
+            type: u4
+          - id: object_table_ptr
+            -orig-id: lpObjectTable
+            type: resolved_ptr
+          - id: reserved
+            -orig-id: dwReserved
+            type: u4
+          - id: unused
+            -orig-id: dwUnused
+            type: u4
+
+          # points to array of pointers to `PrivateObjectDescriptor`
+          - id: object_list_ptr
+            -orig-id: lpObjectList
+            type: resolved_ptr
+          - id: unused_2
+            -orig-id: dwUnused2
+            type: u4
+          - id: sz_project_description
+            -orig-id: szProjectDescription
+            type: u4
+          - id: sz_project_help_file
+            -orig-id: szProjectHelpFile
+            type: u4
+          - id: reserved_2
+            -orig-id: dwReserved2
+            type: u4
+          - id: help_context_id
+            -orig-id: dwHelpContextId
+            type: u4
+
+        instances:
+          object_table:
+            io: _root._io
+            pos: object_table_ptr.ofs
+            type: object_table
+
+
+
+
+  resolved_ptr:
+    seq:
+      - id: ptr
+        type: u4
+    instances:
+      valid:
+        value: ptr != 0
+      resolved:
+        pos: 0
+        type: "resolve_offset(ptr, _root)" # otherwise it won't be passed
+      ofs:
+        value: resolved.ofs.as<u4>
```

Comment:
That unfortunately won't work. Opaque types (such as `resolve_offset`) are called _opaque_, because you can't see what's inside, it's just a black box (from the perspective of KSY). Therefore you can't reference its properties from the Kaitai Struct language, because the compiler is unable to derive the type of the property or check if the property even exists. It makes sense, the opaque type is implemented only in the target language, the compiler hasn't access to the it while compiling KSY and even if it had, it couldn't understand it. _(I see it's unfortunately quite bad for your spec because you're counting on it in many places.)_

You may suggest _"so just turn off the property checking for the opaque types and trust the KSY dev"_, but then the type of the property would be just `AnyType`, and though you could in theory cast it to the arbitrary type you want, I'm still not sure it's wise. It's a quick solution though.

But I think the right thing to do is not to be lazy and think about the [interfaces](https://github.com/kaitai-io/kaitai_struct/issues/314), which would solve it quite nicely. You could define an interface in the KSY, where you specify an `ofs` property of type `u4` and then tell the KSC that the opaque type `resolve_offset` implements this interface.

And it's much more general than just turning off the checks on opaque types, it would also make the KSY development much more convenient. For example, you could have several versions of the same file format in different types and declare that all implement one interface. Then you'd switch them in some huge type `switch`, which makes accessing the common properties a nightmare, but not in that case, because the combined type would be our common interface.

Reply:
>That unfortunately won't work. Opaque types (such as resolve_offset) are called opaque, because you can't see what's inside, it's just a black box (from the perspective of KSY).

For python this already works. For statically typed languages we strictly need interfaces proposal. Unfortunately noone has implemented it and I don't want to touch to Scala code.